File: red/src/cluster.ts
import type {Opts} from 'red/workflow';
import {collect,expand,deployment_requests,plan_deployment,source_cidrs} from 'colors-compute-red';
export const defaultComputeProvider='vultr';
export const defaultNodeCount=3;
export const topology=(opts:Opts)=>[{role:null,count:opts['automq-node-count']??defaultNodeCount}];
export function requirements(opts:Opts) {
 const sources=(name:string)=>source_cidrs(opts,name,'automq-'+name);
 const ingress:any[]=[{id:'ssh',protocol:'tcp',from_port:22,to_port:22,sources:sources('ssh-sources')}];
 const kafka=sources('kafka-sources');if(kafka.length)ingress.push({id:'kafka',protocol:'tcp',from_port:kafkaPort(opts),to_port:kafkaPort(opts),sources:kafka});
 for(const [id,port] of [['controller',controllerPort(opts)],['internal',internalPort(opts)]])ingress.push({id,protocol:'tcp',from_port:port,to_port:port,sources:['private']});
 return {security:{ingress,egress:'all',private_filter:true},private:true,legacy_state_keys:[opts.profile+'/automq-infrastructure.tfstate']};
}
const requests=(opts:Opts)=>deployment_requests(opts,topology(opts),requirements(opts),{mode:'managed',public_key:'ssh-ed25519 PLACEHOLDER managed-by-colors'});
export const nodeCount=(opts:Opts):number=>topology(opts)[0].count;
export const indexes=(opts:Opts):number[]=>expand(topology(opts)).map(n=>n.index);
export const brokerName=(opts:Opts,i:number)=>`${opts['automq-broker-name-prefix']||'b'}${i}.${opts['automq-host']}`;
export const brokerNames=(opts:Opts)=>indexes(opts).map(i=>brokerName(opts,i));
export const certificateNames=(opts:Opts)=>[String(opts['automq-host']),...brokerNames(opts)];
export const computeName=(opts:Opts)=>requests(opts).shared.name;
export const machineName=(opts:Opts,i:number)=>requests(opts).nodes[i].name;
export const machineNames=(opts:Opts)=>indexes(opts).map(i=>machineName(opts,i));
export interface Node {role:string|null;index:number;name:string;ip:string;'vpc-ip':string;user:string;sudoer:string;'broker-name':string;[extra:string]:any}
function automqNode(opts:Opts,node:any):Node{const {vpc_ip,...rest}=node;return {...rest,'vpc-ip':vpc_ip,'broker-name':opts['provider-dns']==='none'?node.ip:brokerName(opts,node.index)};}
export const fallbackNodes=(opts:Opts):Node[]=>plan_deployment(opts,topology(opts),requirements(opts)).cluster.nodes.map(n=>automqNode(opts,n));
export function nodes(opts:Opts,params?:any):Node[]{
 const recorded=params??opts['colors-compute/cluster'];if(!recorded){if(opts['red/event']==='build'||opts['red/dry-run'])return fallbackNodes(opts);throw Error('compute cluster unavailable; refusing placeholder inventory');}
 const declarations=opts['red/event']==='delete'?recorded.nodes:expand(topology(opts));
 const requests=declarations.map((n:any)=>({...n,private:true,provider:opts['provider-compute']}));
 return collect(requests,recorded.nodes,requests[0].node_id).nodes.map(n=>automqNode(opts,n));
}
// ----------------------------------------------------------------- listeners

export function controllerPort(opts: Opts): number {
  return (opts["automq-controller-port"] as number) ?? 9093;
}

export function internalPort(opts: Opts): number {
  return (opts["automq-internal-port"] as number) ?? 9094;
}

export function kafkaPort(opts: Opts): number {
  return (opts["automq-kafka-port"] as number) ?? 9092;
}

// `controller.quorum.voters`, identical on every node.
//
// Static rather than dynamic: three fixed nodes are desired state, and a static
// list is what makes the rendered configuration deterministic and the goldens
// meaningful. Built from VPC addresses — the quorum never crosses the public
// interface.
export function quorumVoters(opts: Opts, list: Node[]): string {
  return list.map((n) => `${n.index}@${n["vpc-ip"]}:${controllerPort(opts)}`).join(",");
}

// `listeners` for node `n`. CONTROLLER and INTERNAL bind the VPC address
// specifically, which is why the container runs with host networking: a bridged
// container cannot bind an address that belongs only to the host. EXTERNAL
// binds every interface because it is the public endpoint.
export function listeners(opts: Opts, n: Node): string {
  return `CONTROLLER://${n["vpc-ip"]}:${controllerPort(opts)}` +
    `,INTERNAL://${n["vpc-ip"]}:${internalPort(opts)}` +
    `,EXTERNAL://0.0.0.0:${kafkaPort(opts)}`;
}

// What node `n` tells clients to come back to. INTERNAL advertises the VPC
// address; EXTERNAL advertises this broker's own public name, which must
// resolve and must be in its certificate. CONTROLLER is deliberately absent —
// Kafka rejects a controller entry in `advertised.listeners`.
export function advertisedListeners(opts: Opts, n: Node): string {
  return `INTERNAL://${n["vpc-ip"]}:${internalPort(opts)}` +
    `,EXTERNAL://${n["broker-name"]}:${kafkaPort(opts)}`;
}

// ---------------------------------------------------------------- principals

function principal(value: unknown, fallback: string): string {
  const text = String(value ?? "");
  return text.length > 0 ? text : fallback;
}

export const adminUser = (opts: Opts) => principal(opts["automq-admin-user"], "automq-admin");
export const brokerUser = (opts: Opts) => principal(opts["automq-broker-user"], "automq-broker");
export const controllerUser = (opts: Opts) =>
  principal(opts["automq-controller-user"], "automq-controller");
export const clientUser = (opts: Opts) => principal(opts["automq-sasl-user"], "automq");

// The principals bootstrapped into the metadata log by the genesis format.
//
// The controller principal is deliberately absent: it authenticates with PLAIN
// from a static JAAS file, precisely so that forming the controller quorum
// depends on nothing stored in the metadata log the quorum is trying to serve.
export function scramPrincipals(opts: Opts): string[] {
  return [adminUser(opts), brokerUser(opts), clientUser(opts)];
}

// `super.users`. The client principal is never here — it is ACL-scoped, and a
// public endpoint whose only authenticated identity is a superuser is an
// authorization hole with a password on it.
export function superUsers(opts: Opts): string {
  return [adminUser(opts), brokerUser(opts), controllerUser(opts)]
    .map((user) => `User:${user}`).join(";");
}

export function topicPrefix(opts: Opts): string {
  return principal(opts["automq-client-topic-prefix"], "colors-");
}

export interface Acl {
  principal: string;
  "resource-type": string;
  "pattern-type": string;
  name: string;
  operations: string[];
}

// The client principal's complete authority, enumerated so it can be read and
// tested rather than inferred. No Create, no Alter, no ClusterAction, no
// TransactionalId — acceptance asserts the denials as well as the grants.
export function clientAcls(opts: Opts): Acl[] {
  const user = clientUser(opts);
  const prefix = topicPrefix(opts);
  return [
    { principal: user, "resource-type": "topic", "pattern-type": "prefixed",
      name: prefix, operations: ["Describe", "Read", "Write"] },
    { principal: user, "resource-type": "group", "pattern-type": "prefixed",
      name: prefix, operations: ["Describe", "Read"] },
  ];
}

File: blue/src/package_automq_blue/cluster.py
"""AutoMQ application facts built from the shared compute contract."""

from __future__ import annotations

from colors_compute import collect, expand
from colors_compute.deployment_request import deployment_requests, source_cidrs
from colors_compute.planning import plan_deployment

DEFAULT_NODE_COUNT = 3
default_compute_provider = 'vultr'


def topology(opts):
    return [{'role': None, 'count': opts.get('automq-node-count', DEFAULT_NODE_COUNT)}]


def _sources(opts, name):
    return source_cidrs(opts, name, 'automq-' + name)


def requirements(opts):
    ingress = [{'id': 'ssh', 'protocol': 'tcp', 'from_port': 22, 'to_port': 22, 'sources': _sources(opts, 'ssh-sources')}]
    kafka = _sources(opts, 'kafka-sources')
    if kafka:
        ingress.append({'id': 'kafka', 'protocol': 'tcp', 'from_port': kafka_port(opts), 'to_port': kafka_port(opts), 'sources': kafka})
    for name, port in [('controller', controller_port(opts)), ('internal', internal_port(opts))]:
        ingress.append({'id': name, 'protocol': 'tcp', 'from_port': port, 'to_port': port, 'sources': ['private']})
    return {'security': {'ingress': ingress, 'egress': 'all', 'private_filter': True},
            'private': True, 'legacy_state_keys': [str(opts.get('profile')) + '/automq-infrastructure.tfstate']}


def _requests(opts):
    return deployment_requests(opts, topology(opts), requirements(opts), {'mode': 'managed', 'public_key': 'ssh-ed25519 PLACEHOLDER managed-by-colors'})


def _s(value) -> str:
    """Clojure's `str`: nil renders empty, booleans lowercase."""
    if value is None:
        return ""
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


# ------------------------------------------------------------------- names


def node_count(opts: dict) -> int:
    """Use normalized library results for application rendering."""
    return topology(opts)[0]['count']


def indexes(opts: dict) -> list[int]:
    """Use normalized library results for application rendering."""
    return [node["index"] for node in expand(topology(opts))]


def broker_name(opts: dict, i: int) -> str:
    """The public name broker ``i`` advertises, ``b<i>.<automq-host>``.

    Kafka redirects a client from the bootstrap name to whatever a broker
    advertises, so this name must resolve publicly and must appear in that
    broker's certificate. Both the DNS stage and the SAN list below derive from
    this one function."""
    prefix = _s(opts.get("automq-broker-name-prefix")) or "b"
    return f"{prefix}{i}.{_s(opts.get('automq-host'))}"


def broker_names(opts: dict) -> list[str]:
    return [broker_name(opts, i) for i in indexes(opts)]


def certificate_names(opts: dict) -> list[str]:
    """The exact SAN list: the bootstrap name plus every broker name.

    Derived rather than guessed. An earlier design used a wildcard, which
    required deriving the zone from the host and left the apex needing its own
    SAN anyway; enumerating the names this cluster actually serves is both
    shorter and checkable."""
    return [_s(opts.get("automq-host")), *broker_names(opts)]


def compute_name(opts: dict) -> str:
    """Use normalized library results for application rendering."""
    return _requests(opts)['shared']['name']


def machine_name(opts: dict, i: int) -> str:
    """The label of machine ``i``, ``<compute-name>-<i>``: the Cluster
    Standard's fallback name for the None role, which is also what the template
    labels the instance. Numbered because there is more than one; the standard
    names the machine after the profile, and the index disambiguates without
    introducing a second naming scheme."""
    return _requests(opts)['nodes'][i]['name']


def machine_names(opts: dict) -> list[str]:
    return [machine_name(opts, i) for i in indexes(opts)]


# --------------------------------------------------------------------- nodes


def _automq_node(opts: dict, node: dict) -> dict:
    """Use normalized library results for application rendering."""
    result = {k: v for k, v in node.items() if k != "vpc_ip"}
    result["vpc-ip"] = node.get("vpc_ip")
    result["broker-name"] = broker_name(opts, node["index"])
    return result


def fallback_nodes(opts: dict) -> list[dict]:
    """Use normalized library results for application rendering."""
    return [_automq_node(opts, n) for n in plan_deployment(opts, topology(opts), requirements(opts))['cluster']['nodes']]


def nodes(opts: dict, params=None) -> list[dict]:
    """Use normalized library results for application rendering."""
    recorded = params or opts.get('colors-compute/cluster')
    if recorded is None:
        if opts.get('blue/event') == 'build' or opts.get('blue/dry-run'):
            return fallback_nodes(opts)
        raise ValueError('compute cluster unavailable; refusing placeholder inventory')
    declarations = recorded.get('nodes', []) if opts.get('blue/event') == 'delete' else expand(topology(opts))
    requests = [{**node, 'private': True, 'provider': opts['provider-compute']} for node in declarations]
    checked = collect(requests, recorded.get('nodes', []), requests[0]['node_id'])
    return [_automq_node(opts, n) for n in checked['nodes']]


# ----------------------------------------------------------------- listeners


def controller_port(opts: dict) -> int:
    value = opts.get("automq-controller-port")
    return 9093 if value is None else value


def internal_port(opts: dict) -> int:
    value = opts.get("automq-internal-port")
    return 9094 if value is None else value


def kafka_port(opts: dict) -> int:
    value = opts.get("automq-kafka-port")
    return 9092 if value is None else value


def quorum_voters(opts: dict, nodes_: list[dict]) -> str:
    """``controller.quorum.voters``, identical on every node.

    Static rather than dynamic: three fixed nodes are desired state, and a
    static list is what makes the rendered configuration deterministic and the
    goldens meaningful. Built from VPC addresses — the quorum never crosses the
    public interface."""
    return ",".join(f"{n['index']}@{n['vpc-ip']}:{controller_port(opts)}"
                    for n in nodes_)


def listeners(opts: dict, n: dict) -> str:
    """``listeners`` for node ``n``. CONTROLLER and INTERNAL bind the VPC
    address specifically, which is why the container runs with host networking:
    a bridged container cannot bind an address that belongs only to the host.
    EXTERNAL binds every interface because it is the public endpoint."""
    return (f"CONTROLLER://{n['vpc-ip']}:{controller_port(opts)}"
            f",INTERNAL://{n['vpc-ip']}:{internal_port(opts)}"
            f",EXTERNAL://0.0.0.0:{kafka_port(opts)}")


def advertised_listeners(opts: dict, n: dict) -> str:
    """What node ``n`` tells clients to come back to. INTERNAL advertises the
    VPC address; EXTERNAL advertises this broker's own public name, which must
    resolve and must be in its certificate. CONTROLLER is deliberately absent —
    Kafka rejects a controller entry in ``advertised.listeners``."""
    return (f"INTERNAL://{n['vpc-ip']}:{internal_port(opts)}"
            f",EXTERNAL://{n['broker-name']}:{kafka_port(opts)}")


# ---------------------------------------------------------------- principals


def _principal(value, fallback: str) -> str:
    return _s(value) or fallback


def admin_user(opts: dict) -> str:
    return _principal(opts.get("automq-admin-user"), "automq-admin")


def broker_user(opts: dict) -> str:
    return _principal(opts.get("automq-broker-user"), "automq-broker")


def controller_user(opts: dict) -> str:
    return _principal(opts.get("automq-controller-user"), "automq-controller")


def client_user(opts: dict) -> str:
    return _principal(opts.get("automq-sasl-user"), "automq")


def scram_principals(opts: dict) -> list[str]:
    """The principals bootstrapped into the metadata log by the genesis format.

    The controller principal is deliberately absent: it authenticates with
    PLAIN from a static JAAS file, precisely so that forming the controller
    quorum depends on nothing stored in the metadata log the quorum is trying to
    serve."""
    return [admin_user(opts), broker_user(opts), client_user(opts)]


def super_users(opts: dict) -> str:
    """``super.users``. The client principal is never here — it is ACL-scoped,
    and a public endpoint whose only authenticated identity is a superuser is an
    authorization hole with a password on it."""
    return ";".join(f"User:{user}" for user in
                    (admin_user(opts), broker_user(opts), controller_user(opts)))


def topic_prefix(opts: dict) -> str:
    return _principal(opts.get("automq-client-topic-prefix"), "colors-")


def client_acls(opts: dict) -> list[dict]:
    """The client principal's complete authority, enumerated so it can be read
    and tested rather than inferred. No Create, no Alter, no ClusterAction, no
    TransactionalId — acceptance asserts the denials as well as the grants."""
    user, prefix = client_user(opts), topic_prefix(opts)
    return [{"principal": user, "resource-type": "topic",
             "pattern-type": "prefixed", "name": prefix,
             "operations": ["Describe", "Read", "Write"]},
            {"principal": user, "resource-type": "group",
             "pattern-type": "prefixed", "name": prefix,
             "operations": ["Describe", "Read"]}]
